src/openrct2/actions/LargeSceneryRemoveAction.hpp

```diff
@@ -144,25 +144,25 @@ DEFINE_GAME_ACTION(LargeSceneryRemoveAction, GAME_COMMAND_REMOVE_LARGE_SCENERY,
 
         rct_scenery_entry* scenery_entry = tileElement->AsLargeScenery()->GetEntry();
 
-        LocationXYZ16 firstTile = { scenery_entry->large_scenery.tiles[_tileIndex].x_offset,
-                                    scenery_entry->large_scenery.tiles[_tileIndex].y_offset,
-                                    static_cast<int16_t>((_loc.z) - scenery_entry->large_scenery.tiles[_tileIndex].z_offset) };
+        CoordsXYZ firstTile = { scenery_entry->large_scenery.tiles[_tileIndex].x_offset,
+                                scenery_entry->large_scenery.tiles[_tileIndex].y_offset,
+                                _loc.z - scenery_entry->large_scenery.tiles[_tileIndex].z_offset };
 
-        rotate_map_coordinates(&firstTile.x, &firstTile.y, _loc.direction);
+        CoordsXY rotatedFirstTile = firstTile.Rotate(_loc.direction);
 
-        firstTile.x = _loc.x - firstTile.x;
-        firstTile.y = _loc.y - firstTile.y;
+        firstTile.x = _loc.x - rotatedFirstTile.x;
+        firstTile.y = _loc.y - rotatedFirstTile.y;
 
         for (int32_t i = 0; scenery_entry->large_scenery.tiles[i].x_offset != -1; i++)
         {
-            LocationXYZ16 currentTile = { scenery_entry->large_scenery.tiles[i].x_offset,
-                                          scenery_entry->large_scenery.tiles[i].y_offset,
-                                          scenery_entry->large_scenery.tiles[i].z_offset };
+            CoordsXYZ currentTile = { scenery_entry->large_scenery.tiles[i].x_offset,
+                                      scenery_entry->large_scenery.tiles[i].y_offset,
+                                      scenery_entry->large_scenery.tiles[i].z_offset };
 
-            rotate_map_coordinates(&currentTile.x, &currentTile.y, _loc.direction);
+            CoordsXY rotatedCurrentTile = currentTile.Rotate(_loc.direction);
 
-            currentTile.x += firstTile.x;
-            currentTile.y += firstTile.y;
+            currentTile.x = rotatedCurrentTile.x + firstTile.x;
+            currentTile.y = rotatedCurrentTile.y + firstTile.y;
             currentTile.z += firstTile.z;
```

Comment:
I found these `+= z` weird in some places, as rotate doesn't touch `z`. Do they sound normal?

Reply:
What this function is doing is deleting a multi tile scenery item:

_loc represents where the user has clicked on the multi tile scenery item.

firstTile is the first tile of the multi tile scenery item. To work out where the first tile is you need to know the offset from the selected tile and go backwards to the first tile. So thats why you rotate the xy offsets as per the user clicked tile. z offsets are obviously not changed by rotation. Then you take it away from _loc to get the firstTile.

After you know the first tile you iterate through each of the remaining tiles by adding on the offsets (rotated xy) of each tile to the first tile.  

So yes the += z is required as the z coordinate of the multi tile element can change and is not related to rotation.

Reply:
I see, thanks for the very detailed explanation!